src/propagate_constant.cpp

```diff
@@ -20,33 +21,42 @@ bool skip_propogate(instruction_ref ins)
     return false;
 }
 
+bool is_const(instruction_ref ins) { return ins->can_eval() and not skip_propogate(ins); }
+
 void propagate_constant::apply(module& m) const
 {
+    std::unordered_set<instruction_ref> const_instrs;
+
+    // Find instructions that can be evaluated to a literal
     for(auto i : iterator_for(m))
     {
-        if(i->name() != "@literal")
-            continue;
-        if(i->outputs().empty())
-            continue;
-        fix([&](auto self, auto ins) {
-            std::unordered_set<instruction_ref> children(ins->outputs().begin(),
-                                                         ins->outputs().end());
-            for(auto child : children)
-            {
-                if(child->name() == "@literal" or skip_propogate(child))
-                {
-                    self(child);
-                    continue;
-                }
-                auto r = child->eval();
-                if(not r.empty())
+        if(not is_const(i) or i == std::prev(m.end()))
```

Comment:
Assign `auto last = std::prev(m.end())` outside of the loop, and then you can just compare `last`.

Also, use early return instead of nesting:

```cpp
for(auto i : iterator_for(m))
{
    if (is_const(i) and i != last)
        continue;
    ...
}
```